test/k8sT/Policies.go

```diff
@@ -808,3 +815,284 @@ var _ = Describe("K8sValidatedPolicyTestAcrossNamespaces", func() {
 
 })
+
+// Create a server pod with a listening container for each port in ports[].
+// Will also assign a pod label with key: "pod-name" and label set to the given podname for later use by the network
+// policy.
+func createServerPodAndService(k *helpers.Kubectl, namespace, podName string, ports []int) (*v1.Pod, *v1.Service) {
+	// Because we have a variable amount of ports, we'll first loop through and generate our Containers for our pod,
+	// and ServicePorts.for our Service.
+	containers := []v1.Container{}
+	servicePorts := []v1.ServicePort{}
+	for _, port := range ports {
+		// Build the containers for the server pod.
+		containers = append(containers, v1.Container{
+			Name:  fmt.Sprintf("%s-container-%d", podName, port),
+			Image: "porter:1.0",
+			Env: []v1.EnvVar{
+				{
+					Name:  fmt.Sprintf("SERVE_PORT_%d", port),
+					Value: "foo",
+				},
+			},
+			Ports: []v1.ContainerPort{
+				{
+					ContainerPort: int32(port),
+					Name:          fmt.Sprintf("serve-%d", port),
+				},
+			},
+			ReadinessProbe: &v1.Probe{
+				Handler: v1.Handler{
+					HTTPGet: &v1.HTTPGetAction{
+						Path: "/",
+						Port: intstr.IntOrString{
+							IntVal: int32(port),
+						},
+						Scheme: v1.URISchemeHTTP,
+					},
+				},
+			},
+		})
+
+		// Build the Service Ports for the service.
+		servicePorts = append(servicePorts, v1.ServicePort{
+			Name:       fmt.Sprintf("%s-%d", podName, port),
+			Port:       int32(port),
+			TargetPort: intstr.FromInt(port),
+		})
+	}
+
+	By(fmt.Sprintf("Creating a server pod %s in namespace %s", podName, namespace))
+	pod, err := k.CoreV1().Pods(namespace).Create(&v1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: podName,
+			Labels: map[string]string{
+				"pod-name": podName,
+			},
+		},
+		Spec: v1.PodSpec{
+			Containers:    containers,
+			RestartPolicy: v1.RestartPolicyNever,
+		},
+	})
+	Expect(err).NotTo(HaveOccurred())
+
+	svcName := fmt.Sprintf("svc-%s", podName)
+	By(fmt.Sprintf("Creating a service %s for pod %s in namespace %s", svcName, podName, namespace))
+	svc, err := k.CoreV1().Services(namespace).Create(&v1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: svcName,
+		},
+		Spec: v1.ServiceSpec{
+			Ports: servicePorts,
+			Selector: map[string]string{
+				"pod-name": podName,
+			},
+		},
+	})
+	Expect(err).NotTo(HaveOccurred())
+
+	return pod, svc
+}
+
+func cleanupServerPodAndService(k *helpers.Kubectl, pod *v1.Pod, service *v1.Service) {
+	By("Cleaning up the server.")
+	err := k.CoreV1().Pods(pod.Namespace).Delete(pod.Name, nil)
+	Expect(err).To(BeNil(), "Terminating containers are not deleted after timeout")
+
+	By("Cleaning up the server's service.")
+	err = k.CoreV1().Services(service.Namespace).Delete(service.Name, nil)
+	Expect(err).To(BeNil(), "Terminating containers are not deleted after timeout")
+}
+
+func createNetworkClientPod(k *helpers.Kubectl, ns, podName string, targetService *v1.Service, dPort int) *v1.Pod {
+	pod, err := k.CoreV1().Pods(ns).Create(&v1.Pod{
```

Comment:
Will the pod not have been created if an error is returned here? I ask because since this function has an `Expect` inside of it, I am worried about cruft being left behind in the test env and affecting other tests, see: https://github.com/cilium/cilium/pull/2938/commits/88e9f85d76188398e0f41b7460d4182245d25cec#diff-3e28e952665f25f47cc466d4d5c9a84cR968

Reply:
>Will the pod not have been created if an error is returned here? 

If there's an error there it can mean a pod is already running with that name for example. Should he delete it? What happens if the developer was running a really important pod with that name?

In case there was no pod running but the pod failed to be created we can assume the pod was not created